[ISSUE] Failures for Auto-Enabled System Tables

While using Terraform with Databricks, we noticed state changes related to system tables during the terraform plan step. The following system tables are affected:

- databricks_system_schema.access_logs is tainted and requires replacement.
- databricks_system_schema.billing_logs is marked for creation.
- databricks_system_schema.compute_logs is tainted and requires replacement.
- databricks_system_schema.lakeflow_logs is tainted and requires replacement.

However, during the `terraform apply` step, encounter the following error, which prevents the deployment of new changes:

```
Error: cannot create system schema: billing system schema can only be enabled by Databricks.

 with databricks_system_schema.billing_logs,
 on main.tf line 226, in resource "databricks_system_schema" "billing_logs":
 226: resource "databricks_system_schema" "billing_logs" {

Error: Process completed with exit code 1.
```

### Root Cause:

The `billing` schema is now enabled by default in Databricks, meaning that any resource block attempting to create it manually will result in an error.

### Proposed Solution:

Documentation Update: Clearly state in the documentation that certain system schemas (such as `billing`) may be auto-enabled and should not be manually declared in Terraform configurations.

Error Handling Improvement: Instead of failing the terraform apply process, introduce logic that detects and ignores errors when certain system tables are already enabled by default.

Terraform Provider Enhancement: Modify the Databricks Terraform provider to automatically detect auto-enabled system schemas and exclude them from the state file or ignore changes during `terraform apply`.

--- a/NEXT_CHANGELOG.md
+++ b/NEXT_CHANGELOG.md
@@ -6,6 +6,8 @@
 
 ### Bug Fixes
 
+* Handle auto-enabled errors with `databricks_system_schema` [#4547](https://github.com/databricks/terraform-provider-databricks/pull/4547)
+
 ### Documentation
 
 ### Exporter

--- a/catalog/resource_system_schema.go
+++ b/catalog/resource_system_schema.go
@@ -50,7 +50,8 @@ func ResourceSystemSchema() common.Resource {
 			SchemaName:  new,
 		})
 		//ignore "schema <schema-name> already exists" error
-		if err != nil && !strings.Contains(err.Error(), "already exists") {
+		//ignore "<schema-name> system schema can only be enabled by Databricks" error
+		if err != nil && !strings.Contains(err.Error(), "already exists") && !strings.Contains(err.Error(), "can only be enabled by Databricks") {
 			return err
 		}
 		//disable old schemas if needed

--- a/catalog/resource_system_schema.go
+++ b/catalog/resource_system_schema.go
@@ -21,6 +21,10 @@ func ResourceSystemSchema() common.Resource {
 			Type:     schema.TypeString,
 			Computed: true,
 		}
+		m["auto_enabled"] = &schema.Schema{
+			Type:     schema.TypeBool,
+			Computed: true,
+		}
 		m["state"].Computed = true
 		return m
 	})
@@ -49,13 +53,20 @@ func ResourceSystemSchema() common.Resource {
 			MetastoreId: metastoreSummary.MetastoreId,
 			SchemaName:  new,
 		})
-		//ignore "schema <schema-name> already exists" error
-		//ignore "<schema-name> system schema can only be enabled by Databricks" error
-		if err != nil && !strings.Contains(err.Error(), "already exists") && !strings.Contains(err.Error(), "can only be enabled by Databricks") {
-			return err
+		if err != nil {
+			//ignore "<schema-name> system schema can only be enabled by Databricks" error, also mark it to make delete no-op
+			d.Set("auto_enabled", strings.Contains(err.Error(), "can only be enabled by Databricks"))
+			//ignore "schema <schema-name> already exists" error
+			if !strings.Contains(err.Error(), "already exists") && !strings.Contains(err.Error(), "can only be enabled by Databricks") {
+				return err
+			}
 		}
 		//disable old schemas if needed
 		if old != "" {
+			if d.Get("auto_enabled").(bool) {
+				log.Printf("[WARN] %s is auto enabled, ignoring it", old)
+				return nil
+			}
 			err = w.SystemSchemas.Disable(ctx, catalog.DisableRequest{
 				MetastoreId: metastoreSummary.MetastoreId,
 				SchemaName:  old,
@@ -97,7 +108,8 @@ func ResourceSystemSchema() common.Resource {
 					// only track enabled/legacy schemas
 					if schema.State != catalog.SystemSchemaInfoStateEnableCompleted &&
 						schema.State != catalog.SystemSchemaInfoStateEnableInitialized &&
-						schema.State != catalog.SystemSchemaInfoStateUnavailable {
+						schema.State != catalog.SystemSchemaInfoStateUnavailable &&
+						schema.State != catalog.SystemSchemaInfoStateAvailable {
 						log.Printf("[WARN] %s is not enabled, ignoring it", schemaName)
 						d.SetId("")
 						return nil
@@ -116,6 +128,10 @@ func ResourceSystemSchema() common.Resource {
 			_, schemaName, err := pi.Unpack(d)
 			if err != nil {
 				return err
+			}
+			if d.Get("auto_enabled").(bool) {
+				log.Printf("[WARN] %s is auto enabled, ignoring it", schemaName)
+				return nil
 			}
 			w, err := c.WorkspaceClient()
 			if err != nil {

--- a/catalog/resource_system_schema.go
+++ b/catalog/resource_system_schema.go
@@ -108,8 +108,7 @@ func ResourceSystemSchema() common.Resource {
 					// only track enabled/legacy schemas
 					if schema.State != catalog.SystemSchemaInfoStateEnableCompleted &&
 						schema.State != catalog.SystemSchemaInfoStateEnableInitialized &&
-						schema.State != catalog.SystemSchemaInfoStateUnavailable &&
-						schema.State != catalog.SystemSchemaInfoStateAvailable {
+						schema.State != catalog.SystemSchemaInfoStateUnavailable {
 						log.Printf("[WARN] %s is not enabled, ignoring it", schemaName)
 						d.SetId("")
 						return nil

--- a/catalog/resource_system_schema_test.go
+++ b/catalog/resource_system_schema_test.go
@@ -1,13 +1,16 @@
 package catalog
 
 import (
+	"errors"
 	"net/http"
 	"testing"
 
+	"github.com/databricks/databricks-sdk-go/experimental/mocks"
 	"github.com/databricks/databricks-sdk-go/service/catalog"
 	"github.com/databricks/terraform-provider-databricks/common"
 	"github.com/databricks/terraform-provider-databricks/qa"
 	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/mock"
 )
 
 func TestSystemSchemaCreate(t *testing.T) {
@@ -85,6 +88,40 @@ func TestSystemSchemaCreate_Error(t *testing.T) {
 	}.Apply(t)
 	qa.AssertErrorStartsWith(t, err, "Internal error happened")
 	assert.Equal(t, "", d.Id(), "Id should be empty for error creates")
+}
+
+func TestSystemSchemaCreateAlreadyEnabled(t *testing.T) {
+	qa.ResourceFixture{
+		MockWorkspaceClientFunc: func(w *mocks.MockWorkspaceClient) {
+			w.GetMockMetastoresAPI().EXPECT().Summary(mock.Anything).Return(&catalog.GetMetastoreSummaryResponse{
+				MetastoreId: "abc",
+			}, nil)
+			e := w.GetMockSystemSchemasAPI().EXPECT()
+			e.Enable(mock.Anything, catalog.EnableRequest{
+				MetastoreId: "abc",
+				SchemaName:  "billing",
+			}).Return(errors.New("billing system schema can only be enabled by Databricks"))
+			e.ListByMetastoreId(mock.Anything, "abc").Return(&catalog.ListSystemSchemasResponse{
+				Schemas: []catalog.SystemSchemaInfo{
+					{
+						Schema: "access",
+						State:  catalog.SystemSchemaInfoStateEnableCompleted,
+					},
+					{
+						Schema: "billing",
+						State:  catalog.SystemSchemaInfoStateEnableCompleted,
+					},
+				},
+			}, nil)
+		},
+		Resource: ResourceSystemSchema(),
+		HCL:      `schema = "billing"`,
+		Create:   true,
+	}.ApplyAndExpectData(t, map[string]any{
+		"schema":    "billing",
+		"id":        "abc|billing",
+		"full_name": "system.billing",
+	})
 }
 
 func TestSystemSchemaUpdate(t *testing.T) {

--- a/catalog/system_schema_test.go
+++ b/catalog/system_schema_test.go
@@ -1,20 +1,19 @@
 package catalog_test
 
 import (
-	"os"
 	"testing"
 
 	"github.com/databricks/terraform-provider-databricks/internal/acceptance"
 )
 
 func TestUcAccResourceSystemSchema(t *testing.T) {
-	if os.Getenv("GOOGLE_CREDENTIALS") != "" {
-		t.Skipf("databricks_system_schema resource not available on GCP")
-	}
 	acceptance.UnityWorkspaceLevel(t, acceptance.Step{
 		Template: `
 		resource "databricks_system_schema" "this" {
 			schema = "access"
+		}
+		resource "databricks_system_schema" "this" {
+			schema = "billing"
 		}`,
 	})
 }

--- a/catalog/system_schema_test.go
+++ b/catalog/system_schema_test.go
@@ -9,10 +9,10 @@ import (
 func TestUcAccResourceSystemSchema(t *testing.T) {
 	acceptance.UnityWorkspaceLevel(t, acceptance.Step{
 		Template: `
-		resource "databricks_system_schema" "this" {
+		resource "databricks_system_schema" "access" {
 			schema = "access"
 		}
-		resource "databricks_system_schema" "this" {
+		resource "databricks_system_schema" "billing" {
 			schema = "billing"
 		}`,
 	})

--- a/docs/resources/system_schema.md
+++ b/docs/resources/system_schema.md
@@ -9,6 +9,8 @@ subcategory: "Unity Catalog"
 
 Manages system tables enablement. System tables are a Databricks-hosted analytical store of your account’s operational data. System tables can be used for historical observability across your account. System tables must be enabled by an account admin.
 
+Certain system schemas (such as `billing`) may be auto-enabled once GA and should not be manually declared in Terraform configurations.
+
 ## Example Usage
 
 Enable the system schema `access`

--- a/docs/resources/system_schema.md
+++ b/docs/resources/system_schema.md
@@ -7,9 +7,9 @@ subcategory: "Unity Catalog"
 
 -> This resource can only be used with a workspace-level provider!
 
-Manages system tables enablement. System tables are a Databricks-hosted analytical store of your account’s operational data. System tables can be used for historical observability across your account. System tables must be enabled by an account admin.
+-> Certain system schemas (such as `billing`) may be auto-enabled once GA and should not be manually declared in Terraform configurations.
 
-Certain system schemas (such as `billing`) may be auto-enabled once GA and should not be manually declared in Terraform configurations.
+Manages system tables enablement. System tables are a Databricks-hosted analytical store of your account’s operational data. System tables can be used for historical observability across your account. System tables must be enabled by an account admin.
 
 ## Example Usage
 

